### ISSUE TITLE
Computing of im_neg_kl

### ISSUE THREAD

**Issue by ZeWang95** [yilundu/improved_contrastive_divergence#1]
Thanks for sharing the code!
I'm studying your paper and implementation and I have a question now that might be stupid.
Between L254 and L294 in train.py, I cannot see the difference beween im_neg_kl and im_neg in the last step.
I believe the energy in L255 should equals to L288, and im_neg_kl should be the same as the im_neg before updating. 
Thus shouldn't the im_grad be identical in L289 and L260?

Thanks!

**Comment by yilundu** [comment]
Hey! Good question! im_grad is indeed identical for L289 and L260 -- the one in L289 has a gradient attached to it though while the one in L260 does not. The reason why there are two different variables for im_neg_kl and im_neg is in the case you wish to only have gradients attached a subset of the image samples (L266-L282).

**Comment by ZeWang95** [comment]
Thanks for the quick response!!!
I see...
What do you mean by 'has a gradient attached to it'?

**Comment by yilundu** [comment]
Im_neg (L260) does note have backpropogation through the gradient computation while im_neg_kl (L289) does have backpropogation through the gradient computation
